src/v/resource_mgmt/memory_sampling.cc

```diff
@@ -0,0 +1,224 @@
+    return "Top-N alloc sites - size count stack:";
+}
+
+/// Put `top_n` allocation sites into the front of `allocation_sites`
+static void top_n_allocation_sites(
+  std::vector<ss::memory::allocation_site>& allocation_sites, size_t top_n) {
+    std::partial_sort(
+      allocation_sites.begin(),
+      allocation_sites.begin() + top_n,
+      allocation_sites.end(),
+      [](const auto& lhs, const auto& rhs) { return lhs.size > rhs.size; });
+}
+
+ss::noncopyable_function<void(ss::memory::memory_diagnostics_writer)>
+memory_sampling::get_oom_diagnostics_callback() {
+    // preallocate those so that we don't allocate on OOM
+    std::vector<seastar::memory::allocation_site> allocation_sites(1000);
+    std::vector<char> format_buf(1000);
+
+    return [allocation_sites = std::move(allocation_sites),
+            format_buf = std::move(format_buf)](
+             seastar::memory::memory_diagnostics_writer writer) mutable {
+        auto num_sites = ss::memory::sampled_memory_profile(allocation_sites);
+
+        const size_t top_n = std::min(size_t(10), num_sites);
+        top_n_allocation_sites(allocation_sites, top_n);
+
+        writer(diagnostics_header());
+        writer("\n");
+
+        for (size_t i = 0; i < top_n; ++i) {
+            auto bytes_written = fmt::format_to_n(
+                                   format_buf.begin(),
+                                   format_buf.size(),
+                                   "{}",
+                                   allocation_sites[i])
+                                   .size;
+
+            writer(std::string_view(format_buf.data(), bytes_written));
+        }
+    };
+}
+
+/// We want to print the top-n allocation sites on OOM
+/// Set this up and make sure we don't allocate extra at that point
+void setup_additional_oom_diagnostics() {
+    seastar::memory::set_additional_diagnostics_producer(
+      memory_sampling::get_oom_diagnostics_callback());
+}
+
+void memory_sampling::start_low_available_memory_logging() {
+    // We want some periodic logging "on the way" to OOM. At the same time we
+    // don't want to spam the logs. Hence, we periodically look at the available
+    // memory low watermark (this is without the batch cache). If we see that we
+    // have crossed the 10% and 20% marks we log the allocation sites. We stop
+    // afterwards.
+    //
+    // This logic could also be integrated into batch_cache::reclaim (so that we
+    // actually run when memory is tight) but that is quite critical code so
+    // better not to add anything not needed there.
+    size_t first_log_limit = _first_log_limit_percent
+                             * seastar::memory::stats().total_memory();
+    size_t second_log_limit = _second_log_limit_percent
+                              * seastar::memory::stats().total_memory();
+    size_t next_log_limit = first_log_limit;
+    size_t last_seen_watermark = std::numeric_limits<size_t>::max();
+
+    _timer.set_callback([this,
```

Comment:
Using a timer doesn't strike me as a good solution. We are constantly checking something which will almost never trigger and furthermore if we don't check it rapidly, we may just run out of memory before the timer fires again.

If we don't want to do the actual logging in reclaim, we should at least set up the logging to run after we return from reclaim. 

Batch cache has an "async" reclaim which doesn't run in the context of a memory allocation, but just when we cross some low water mark so it could go there, or it could be in synchronous relcaim but it only triggers the logging to run after (e.g., via some mechanism like `ss::semaphore`).